File: code/python/295.py
# https://leetcode.com/problems/find-median-from-data-stream/
class MedianFinder:
    def __init__(self):
        self.left = []
        self.right = []
        self.is_even = True

    def addNum(self, num: int) -> None:
        if self.is_even:
            heappush(self.right, num)
            heappush(self.left, -heappop(self.right))
        else:
            heappush(self.left, -num)
            heappush(self.right, -heappop(self.left))
        self.is_even = not self.is_even

    def findMedian(self) -> float:
        if self.is_even:
            return (self.right[0] - self.left[0]) / 2
        else:
            return -self.left[0]

File: code/java/295.java
// https://leetcode.com/problems/find-median-from-data-stream/
class MedianFinder {
    private final PriorityQueue<Integer> left = new PriorityQueue<>((a, b) -> b - a);
    private final PriorityQueue<Integer> right = new PriorityQueue<>();

    public void addNum(int num) {
        if (left.isEmpty() || num <= left.peek())
            left.offer(num);
        else
            right.offer(num);
        if (left.size() > right.size() + 1)
            right.offer(left.poll());
        else if (right.size() > left.size())
            left.offer(right.poll());
    }

    public double findMedian() {
        if (left.size() > right.size()) 
            return left.peek();
        return (left.peek() + right.peek()) / 2.0;
    }
}
